tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -692,608 +505,989 @@ void fillShaderBufferFromLongVectorData(std::vector<BYTE> &ShaderBuffer,
-        using ElementType = typename std::decay_t<decltype(Vec)>::value_type;
-        return this->verifyOutput<ElementType>(TestResult, Vec);
-      },
-      ExpectedVector);
+template <typename T, size_t ARITY>
+InputSets<T, ARITY> buildTestInputs(const TestConfig &Config,
+                                    size_t SizeToTest) {
+  InputSets<T, ARITY> Inputs;
+  for (size_t I = 0; I < ARITY; ++I) {
+
+    uint16_t OperandScalarFlag = 1 << I;
+    bool IsOperandScalar = Config.ScalarInputFlags & OperandScalarFlag;
+
+    if (Config.InputValueSetNames[I].IsEmpty()) {
+      LOG_ERROR_FMT_THROW(
+          L"Expected parameter 'InputValueSetName%d' not found.", I + 1);
+      continue;
+    }
+
+    Inputs[I] = buildTestInput<T>(Config.InputValueSetNames[I],
+                                  IsOperandScalar ? 1 : SizeToTest);
+  }
+
+  return Inputs;
 }
 
-// Private version of verifyOutput. Called by the public version of verifyOutput
-// which resolves OutT based on the ExpectedVector type. Most intrinsics will
-// have an OutT that matches the input type being tested (which is T). But some,
-// such as the 'AsType' ops, i.e 'AsUint' have an OutT that doesn't match T.
-template <typename T>    // Primary template from TestConfig
-template <typename OutT> // Secondary template for verifyOutput
-bool TestConfig<T>::verifyOutput(
-    const std::shared_ptr<st::ShaderOpTestResult> &TestResult,
-    const std::vector<OutT> &ExpectedVector) {
+struct ValidationConfig {
+  float Tolerance = 0.0f;
+  ValidationType Type = ValidationType_Epsilon;
 
-  WEX::Logging::Log::Comment(WEX::Common::String().Format(
-      L"verifyOutput with OpType: %ls ExpectedVector<%S>", OpTypeName.c_str(),
-      typeid(OutT).name()));
+  static ValidationConfig Epsilon(float Tolerance) {
+    return ValidationConfig{Tolerance, ValidationType_Epsilon};
+  }
 
-  DXASSERT(!ExpectedVector.empty(),
-           "Programmer Error: ExpectedVector is empty.");
+  static ValidationConfig Ulp(float Tolerance) {
+    return ValidationConfig{Tolerance, ValidationType_Ulp};
+  }
+};
 
-  MappedData ShaderOutData;
-  TestResult->Test->GetReadBackData("OutputVector", &ShaderOutData);
+template <typename T, typename OUT_TYPE, typename OP_TYPE, size_t ARITY>
+void runAndVerify(const TestConfig &Config, OP_TYPE OpType,
+                  const InputSets<T, ARITY> &Inputs,
+                  const std::vector<OUT_TYPE> &Expected,
+                  std::string ExtraDefines,
+                  const ValidationConfig &ValidationConfig) {
 
-  const size_t OutputVectorSize = ExpectedVector.size();
+  std::optional<std::vector<OUT_TYPE>> Actual =
+      runTest<OUT_TYPE>(Config, OpType, Inputs, Expected.size(), ExtraDefines);
 
-  std::vector<OutT> ActualValues;
-  fillLongVectorDataFromShaderBuffer(ShaderOutData, ActualValues,
-                                     OutputVectorSize);
+  if (!Actual) {
+    // The test didn't actually run, so there's nothing to verify.
+    return;
+  }
```

Comment:
nit that I'm never fully sold on:

```suggestion
  // The test didn't actually run, so there's nothing to verify.
  if (!Actual)
    return;
```

Reply:
Can't quite argue against it though....so I changed it.